### PR TITLE
[FIX] l10n_br, base_vat: move Brazilian vat check to base_vat

### PR DESCRIPTION
**Pull request by ushyme**
### Summary
Currently, trying to CPF tax ID on a Brazilian contact always fails.

### Steps to reproduce
* install `contacts` and `base_vat`
* create a new contact with country Brazil
* enter a valid CPF (ex: 11144477735)

You should be met with an error

### Cause
The CPF check is only implemented in `l10n_br`. Without this module, the
default vat check is applied, but it fails for CPF numbers.

### Fix
Move the CPF/CNPJ check from `l10n_br` to `base_vat`

opw-3421476

### CODE REVIEW

File: addons/base_vat/models/res_partner.py [hunk 1]
```diff
@@ -257,7 +258,7 @@ def _build_vat_error_message(self, country_code, wrong_vat, record_label):
             company = self.env.company
 
         vat_label = _("VAT")
-        if country_code and company.country_id and country_code == company.country_id.code.lower():
+        if country_code and company.country_id and country_code == company.country_id.code.lower() and company.country_id.vat_label:
```
Comment (gawa-odoo) [inline addons/base_vat/models/res_partner.py:261]: Is there a reason for this? I mean, you would change the label for all localization, even if it's not a VAT for them, and Tax ID is more suitable for them.
If it's just for Brazil, shouldn't we add it in res_country_data instead?

File: addons/l10n_br/models/res_partner.py [hunk 2]
```diff
@@ -13,67 +13,3 @@ class ResPartner(models.Model):
...
-
-        for partner in self:
-            if not partner.vat:
-                return
-            if not partner.country_code == 'BR':
-                return super().check_vat()
-            vat_clean = list(map(int, re.sub("[^0-9]", "", partner.vat)))
-            if len(vat_clean) == 14:
-                if not _l10n_br_is_valid_cnpj(vat_clean):
-                    raise ValidationError(_("Invalid CNPJ. Make sure that all the digits are entered correctly."))
```
Comment (gawa-odoo) [inline addons/l10n_br/models/res_partner.py:74]: I think you can/should remove the error message from the po/pot. As we're in a localization module, they won't ever be removed. (no transifex)

File: addons/base_vat/models/res_partner.py [hunk 3]
```diff
@@ -704,6 +705,62 @@ def check_vat_t(self, vat):
...
+        Example of a Brazilian CNPJ number: 76.634.583/0001-74.
+        The 13th digit is the check digit of the previous 12 digits.
+        The check digit is calculated by multiplying the first 12 digits by weights and calculate modulo 11 of the result.
+        The 14th digit is the check digit of the previous 13 digits. Calculated the same way.
+        Both remainders are appended to the first 12 digits.
+        '''
+        if not vat:
+            return False
+
+        def _l10n_br_calculate_mod_11(check, weights):
```
Comment (gawa-odoo) [inline addons/base_vat/models/res_partner.py:719]: I'm not sure we want to keep the prefix l10n_br (if really needed, we could keep br if we fear conflict name in function, but not l10n at least)

File: addons/l10n_br/models/res_partner.py [hunk 4]
```diff
@@ -13,67 +13,3 @@ class ResPartner(models.Model):
...
-            # Check the second verification digit
-            if remainder != vat_clean[10]:
-                return False
-            return True
-
-        for partner in self:
-            if not partner.vat:
-                return
-            if not partner.country_code == 'BR':
-                return super().check_vat()
```
Comment (gawa-odoo) [inline addons/l10n_br/models/res_partner.py:70]: Glad you're removing this, nice loops :smile:

File: addons/base_vat/models/res_partner.py [hunk 5]
```diff
@@ -704,6 +705,62 @@ def check_vat_t(self, vat):
...
+            # If the remainder is 10 or 11, set it to 0
+            if remainder in (10, 11):
+                remainder = 0
+            # Check the second verification digit
+            if remainder != vat_clean[10]:
+                return False
+            return True
+
+        vat_clean = list(map(int, re.sub("[^0-9]", "", vat)))
+        return len(vat_clean) == 14 and _l10n_br_is_valid_cnpj(vat_clean) or \
```
Comment (gawa-odoo) [inline addons/base_vat/models/res_partner.py:761]: either one-line it, or put parenthesis around it, so you don't need the `\`

**Review by gawa-odoo — CHANGES_REQUESTED** [review]
Hello, thanks for the fixes.
Some small remarks.
Also could you change/add another number that is valid? The one you give in the commit/PR message is fine (it's the ones with 11 digits that falsely raise an error, it seems).

**Comment by ushyme** [comment]
Oops, I copied the wrong number indeed. It should be correct now.
Thanks for the review!

File: addons/base_vat/models/res_partner.py (hunk 1 above)
Comment (ushyme) [reply addons/base_vat/models/res_partner.py:261]: This is a bit of an unrelated change. Basically, If the new user has the same country as the current company, we replace 'VAT' by the country's actual `vat_label`. However, if that country doesn't have a `vat_label` defined, `False` is displayed instead.

Comment (gawa-odoo) [reply addons/base_vat/models/res_partner.py:261]: Indeed sorry, I misread the message

**Review by gawa-odoo — APPROVED** [review]
LGTM!
Thanks for the changes

**Comment by dpi-odoo** [comment]
Hello!

image

I was testing your PR in this DB (https://52249518-16-0-all.runbot175.odoo.com/web#id=66&cids=1&menu_id=135&action=179&model=res.partner&view_type=form) and I wasn't able to save me CPF on a Contact.

I've attached the screenshot to the comment.

My CPF number is 71882684109, if you wish to test with a real one.

Thank you for looking into this task.

Best,

**Comment by ushyme** [comment]
Hello @dpi-odoo, 

Good catch. 
I somehow introduced a typo when I pushed the last changes.
It's fixed now, thanks !

**Comment by dpi-odoo** [comment]
Hello again!

I hope you had a beautiful weekend,

I just tested your new PR in this DB (https://52249518-16-0-all.runbot175.odoo.com/web#id=66&cids=5&menu_id=135&action=179&model=res.partner&view_type=form) and I wasn't able to save me CPF on a Contact.
image (2)

I'v
image (1)
e attached some other screenshots to the note.

My CPF number is 71882684109 if you wish to test with a real one.

Tests to be run to ensure this is fixed:

CPF Test:

1- Go to the Contact Form View for any record
2- Set the "Country" as Brazil
3- On the "Tax ID" Field input an 11-digit number like "71882684109", Odoo should accept this.
4- On the "Tax ID" Field input an 11-digit number like "718826841-09", Odoo should accept this.
5- On the "Tax ID" Field input a NON 11-digit number. Odoo should not accept this.

CNPJ Test:

1- Go to the Contact Form View for any record
2- Set the "Country" as Brazil
3- On the "Tax ID" Field input a 14-digit number like "49233848000150", Odoo should accept this.
4- On the "Tax ID" Field input a 14-digit number like "49.233.848/0001-50", Odoo should accept this.
5- On the "Tax ID" Field input a NON 14-digit number. Odoo shouldn't accept this.

**Review by FlorianGilbert — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**